IPAddress allocator

aojea opened the pull request:
/kind bug
/kind feature

#### What this PR does / why we need it:

This is the first step towards the implementation of KEP 1880, it introduces a new API object `IPAddress` and a feature gate `MultiCIDRServiceAllocator`

The feature gate enables a new ClusterIP Allocator that uses the IPAddress object instead of the the actual bitmap allocator, removing the current service cidr size limitations (ipv4 is a /12 and ipv6 /108) and setting the maximum size of a service CIDR in /64

The new IPAllocator uses IPAddress objects that reference the Service to store the ClusterIP, creating a 1-to-1 relation.
This relation is maintained by a repair controller that runs in the apiserver.

```release-note
- Added a new IPAddress object kind
- Added a new ClusterIP allocator. The new allocator removes previous Service CIDR block size limitations for IPv4, and limits IPv6 size to a /64
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
Please use the following format for linking documentation:
```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/1880-multiple-service-cidrs
```

aojea committed:
add feature gate

Add MultiCIDRServiceAllocator feature gate.

aojea committed:
hack import aliases

aojea committed:
Add IPAddress API to the REST registry

ipaddress registry storage

don't generate names for ipaddress

aojea committed:
add integration test

aojea committed:
Add IPAddress API

Change-Id: I9cf710f011b58409ab880d3b2e7f841f228ee5ee

aojea committed:
IPAddress validation

Validate IPAddress name is in canonical format
Validate ParentRef is required, and Resource and Name.
Validate IPAddress is inmutable on update.

aojea committed:
integration etcd data

aojea committed:
add apis to apiserver storage

Change-Id: Iea1263ad612c13b93baf8a07641265bf56f08728

aojea committed:
enable IPAddress on storage

Change-Id: Iaeac050223bb7db6acba601aa9b7254c74f2b5d2

aojea committed:
kubectl describe IPAddress

Change-Id: I201c2980830058c92b66e2386bfe8bedd8f0dcd4

aojea committed:
IPAddress printers

Change-Id: I60f40f4fe20b3ebd61b8ee137e55bcdefc1f2c96

diadj commented:
Comment faire pour accéder aux codes de iptv

cici37 commented:
/remove-sig api-machinery

aojea commented on pkg/apis/networking/validation/validation.go:
```diff
@@ -741,3 +742,73 @@ func validateClusterCIDRUpdateSpec(update, old *networking.ClusterCIDRSpec, fldP
+func validateIPAddressParentReference(params *networking.ParentReference, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if params == nil {
+		return allErrs
+	}
```
should we enforce that IP has a reference?

aojea commented on pkg/features/kube_features.go:
```diff
@@ -703,7 +710,7 @@ const (
 
 	// owner: @aojea
-	// kep: https://kep.k8s.io/3070
+	// kep: http://kep.k8s.io/3070
```
this should be https

aojea committed:
add repair loop

Change-Id: I63464bdd5db706ddf7dc5d828b8d03ad532d7981

aojea commented on pkg/apis/networking/types.go:
```diff
@@ -699,3 +700,53 @@ type ClusterCIDRList struct {
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// IPAddress represents an IP used by Kubernetes associated to an IPRange.
+// The name of the object is the IP address in canonical format.
```
examples and reference to the RFC

aojea commented on pkg/apis/networking/validation/validation.go:
```diff
@@ -741,3 +742,73 @@ func validateClusterCIDRUpdateSpec(update, old *networking.ClusterCIDRSpec, fldP
+}
+
+// ValidateIPAddressUpdate tests if an update to an IPAddress is valid.
+func ValidateIPAddressUpdate(update, old *networking.IPAddress) field.ErrorList {
```
TODO, discussion parent reference, understasnd how delete works

aojea commented on pkg/registry/core/service/ipallocator/ipallocator.go:
```diff
@@ -0,0 +1,470 @@
+	if err != nil {
+		return err
+	}
+	addr := netip.MustParseAddr(ip.String())
```
comment about it was already validated

aojea commented on pkg/registry/core/service/ipallocator/ipallocator.go:
```diff
@@ -0,0 +1,470 @@
+		return err
+	}
+	addr := netip.MustParseAddr(ip.String())
+	// check address is within the range of available addresses
```
expand on this comment

aojea committed:
allow to set the service reference on the allocator

aojea committed:
remove apiserver limitations

Change-Id: I97b35d912ba5d86857cf82d3eddd65b648030005

cici37 commented:
/remove-sig api-machinery

cici37 commented:
/remove-sig api-machinery

aojea commented:
/assign @thockin @BenTheElder 

This is ready for review, there is a bunch of commits but most of them are apimachinery. mechanics.

There are 3 importants parts to review:
- ipallocator
- repair controller (I started it as a mimic of the existing one, using a reconcile loop for all the services, but is too slow, so I moved it to the common controller pattern that is much faster and scalable)
- mechanics to plumb the allocator and the repair controller

I'm quite happy with the integration tests , they were very useful during the development to find weird bugs.

aojea commented:
The failures is unrelated, this is good to go now, all tests passing

aojea commented:
/test

aojea commented on pkg/registry/core/service/ipallocator/controller/repairip.go:
```diff
@@ -111,38 +126,99 @@ func NewRepairIPAddress(interval time.Duration,
+				r.ipQueue.Add(key)
+			}
+		},
+	}, interval)
```
This generates a fake update event every interval so all cache elements gets processed again, this is similar to previous repair loop, but I don't have clear we need it here

cici37 commented:
/remove-sig api-machinery

khenidak commented on pkg/apis/networking/types.go:
```diff
@@ -699,3 +700,58 @@ type ClusterCIDRList struct {
+
+// IPAddressSpec describe the attributes in an IP Address,
+type IPAddressSpec struct {
+	// ParentRef references the resources that an IPAddress wants to be attached to.
```
is it `IPAddress` wants to attach to it? or the other way around?

aojea replied:
is the IPAddress parent, we need this to define the relation Service.ClusterIP <--> IPAddress.
We tried to use OwnerReferences on the Service object, the goal was that destroying the parent , it  cascaded destroying the childs ... butds there is a limitation that namespaced objects can not reference non-namespaced objects

khenidak replied:
Then it is 1:1 relationship. a better doc would be something like:

```
ParentRef references the resource that an IPAddress is attached to.
```

aojea replied:
got it

khenidak commented on pkg/apis/networking/types.go:
```diff
@@ -699,3 +700,58 @@ type ClusterCIDRList struct {
+// An IP address can be represented in different formats, to guarantee the uniqueness of the IP,
+// the name of the object is the IP address in canonical format, four decimal digits separated
+// by dots suppressing leading zeros for IPv4 and the representation defined by RFC 5952 for IPv6.
+// Valid: 192.168.1.5 or 2001:db8::1
```
should add the long form ipv6 here as well.

aojea replied:
so, we need the uniqueness on the object to allow to have concurrency on ClusterIP assignment without duplicates, if we allow multiple representations of the same IP we can have Services with duplicate ClusterIPs

khenidak replied:
ack. Valid point.

I just wanted an example of a full IPv6 (without zero octets in the middle) which is still a unique value.

khenidak commented on pkg/apis/networking/validation/validation.go:
```diff
@@ -741,3 +742,75 @@ func validateClusterCIDRUpdateSpec(update, old *networking.ClusterCIDRSpec, fldP
+func ValidateIPAddressName(name string, prefix bool) []string {
+	var errs []string
+	if prefix {
+		errs = append(errs, "prefix not allowed")
```
I really doubt that any user will understand this error message. This is the auto generated name stuff right?

aojea replied:
I have to fix it

aojea replied:
@liggitt I need help here,  I realize that I don't fully understand what this prefix boolean means, I can see the description here https://github.com/kubernetes/kubernetes/blob/561a35f3582abb25da76d31c7c2abc4277054b04/staging/src/k8s.io/apimachinery/pkg/api/validation/generic.go#L29-L34
but I don't know how to treat it for this case that the name can only be a representation of an IP address

liggitt replied:
when `prefix` is true, this is expected to validate the passed-in value is valid as `generateName`

if you don't support generating names, you can validate the same way whether prefix is true or false, and implement GenerateName in your strategy to make the name "generator" for this type just return the base.

khenidak commented on pkg/apis/networking/types.go:
```diff
@@ -699,3 +700,58 @@ type ClusterCIDRList struct {
+	// +required
+	ParentRef *ParentReference
+}
+type ParentReference struct {
```
at some point of time we will have an idiomatic way across the board to do that. here is hoping..

thockin replied:
https://github.com/kubernetes/kubernetes/pull/115075#discussion_r1135721159

khenidak commented on pkg/apis/networking/validation/validation_test.go:
```diff
@@ -2195,3 +2195,199 @@ func TestValidateClusterConfigUpdate(t *testing.T) {
+		},
+	}
+
+	for name, testCase := range testCases {
```
there is no case in which group field is used correctly. is that expected?

aojea replied:
added

khenidak commented on pkg/features/kube_features.go:
```diff
@@ -555,6 +555,13 @@ const (
+	// kep: https://kep.k8s.io/1880
+	// alpha: v1.27
+	//
+	// Allow to dynamically configure Service IP ranges
```
Enables the dynamic configuration of Service IP ranges

khenidak commented on cmd/kube-apiserver/app/options/validation.go:
```diff
@@ -35,7 +36,10 @@ import (
-	const maxCIDRBits = 20
+	maxCIDRBits := 20
+	if utilfeature.DefaultFeatureGate.Enabled(features.MultiCIDRServiceAllocator) {
+		maxCIDRBits = 64
```
How are we planning a revert? what if somebody started using the feature and then decided nope i want to go back to normal?

aojea replied:
yeah, reality is that the Services misconfiguration is broken with and without this feature too https://github.com/kubernetes/kubernetes/issues/114743

This will require manual intervention, use a service-cidr with a valid range, and delete manually the Services that will be OutOfRange. The repair loops will recover the old allocators and raise warnings and events for the Services with clusterIPs out of range.

The funny thing is that they may work, at least with kube-proxy, since kube-proxy doesn't have any notion of the service-cidr, it will install the correct iptables rules.

khenidak replied:
ack - please make sure it is part of the documentation.

khenidak commented on pkg/apis/networking/fuzzer/fuzzer.go:
```diff
@@ -74,5 +76,31 @@ var Funcs = func(codecs runtimeserializer.CodecFactory) []interface{} {
+		bytes[i] = uint8(c.Rand.Intn(255))
+	}
+
+	ip, ok := netip.AddrFromSlice(bytes)
```
I honestly don't think that this will ever fail. It may get a broadcast IP but it will be a valid ip. Also what is the case with empty? shouldn't we try - bounded count of trials?

aojea replied:
I don't know much about how the fuzzer work, I just thought it will be interesting to be totally random

khenidak replied:
Then at least let us try (bounded retry) to get a correct IP, don't return empty string here.

aojea replied:
done

khenidak commented on pkg/controlplane/controller.go:
```diff
@@ -172,16 +176,38 @@ func (c *Controller) Start() {
+			c.informers.Core().V1().Services(),
+			c.informers.Networking().V1alpha1().IPAddresses(),
+		)
+		runRepairClusterIPs = func(stopCh chan struct{}) {
```
This is the same in both legs of the if statement, why not put it outside the if statement?

aojea replied:
because this way I keep the Runner the same and reduce the diff, tomato tomatoe , no?
```
	c.runner = async.NewRunner(c.RunKubernetesNamespaces, c.RunKubernetesService, runRepairNodePorts, runRepairClusterIPs)
```

aojea replied:
See line 213, this doesn't move forward until the repair run once

khenidak replied:
ack.

khenidak commented on pkg/printers/internalversion/printers.go:
```diff
@@ -2779,6 +2787,41 @@ func printClusterCIDRList(list *networking.ClusterCIDRList, options printers.Gen
+		}
+		parentRefName = strings.ToLower(gr.String())
+		if obj.Spec.ParentRef.Namespace != "" {
+			parentRefName += "/" + obj.Spec.ParentRef.Namespace
```
Can an IPAddress be owned by a non namespaced object?

aojea replied:
I never thought about that, main use cases are Services and Gateway API, @thockin WDYT

aojea replied:
I think it should be owned by a non-namespace object, with the new multi-network group APIS I see that some APIs that may require IP addresses are non-namespaced

aojea replied:
I can see that I also implemented it this way, in validation I added a comment that namespace is optional.

This is also allowd, non-namespaced -> non-namespaced and namespaced -> non-namespaced are ok

khenidak commented on pkg/printers/internalversion/printers.go:
```diff
@@ -2779,6 +2787,41 @@ func printClusterCIDRList(list *networking.ClusterCIDRList, options printers.Gen
+	}
+
+	parentRefName := "<none>"
+	if obj.Spec.ParentRef != nil {
```
Validation does not allow a nil value.

aojea replied:
yeah, I like your thoughts on this
I started with ParentRef as optional, but then I thought that  IPAddress is an object to be owned always and for different objects, Services, GatewayAPI, ServiceMesh ...  and allowing to leave dangling IPAddresses will make it hard for implementing multiple controllers ... who reconciles what, and garbage collect, ... having always a reference makes this much simpler

aojea replied:
I just keep this in case we revert the original decisions and ParentRef moves to optional, this is a printer and will not have any other effect

danwinship commented on pkg/registry/core/service/ipallocator/controller/repairip.go:
```diff
@@ -0,0 +1,320 @@
+)
+
+// Repair is a controller loop that periodically examines all service ClusterIP allocations
+// and logs any errors, and then sets the compacted and accurate list of all allocated IPs.
```
I assume this text was copied from the old allocator, but it seems not quite right for the new one? (Especially "compacted".)

aojea replied:
yeah, I have copied it , I need to improve the documentation of this controller

danwinship commented on pkg/registry/core/service/ipallocator/controller/repairip.go:
```diff
@@ -0,0 +1,320 @@
+	client   kubernetes.Interface
+	interval time.Duration
+
+	networkByFamily map[v1.IPFamily]*net.IPNet // networks we operate on, by their family
```
you should use `netutils.IPFamily` here...
(this may require bumping `k8s.io/utils`)

danwinship commented on pkg/registry/core/service/ipallocator/controller/repairip.go:
```diff
@@ -0,0 +1,320 @@
+// * ClusterIPs that do not match the currently configured range
+// * Allocations to services that were not actually created due to a crash or powerloss
+// * Migrates old versions of Kubernetes services into the atomic ipallocator model automatically
+// * ClusterIP out of range cused by a change on the serviceCIDR configured
```
It would be better to separately list the things that it _fixes_ vs the things that it merely logs about

danwinship commented on pkg/registry/core/service/ipallocator/controller/repairip.go:
```diff
@@ -0,0 +1,320 @@
+		}
+	}
+
+	// We have checked that every Service has its corresponding IP
+	// check that there is no IP created by the allocator without
+	// a Service associated.
```
needs a period after "IP" and a capital "C" in "check"

danwinship commented on pkg/registry/core/service/ipallocator/controller/repairip.go:
```diff
@@ -0,0 +1,320 @@
+	}
+	// Check every IPAddress matches the corresponding Service, and rebuild the state as we think it should be.
+	for _, ipAddress := range ipAddresses {
+		// does not reference a Service but is
```
"but is" what?

danwinship commented on pkg/registry/core/service/ipallocator/controller/repairip.go:
```diff
@@ -0,0 +1,320 @@
+		svc, err := r.serviceLister.Services(ipAddress.Spec.ParentRef.Namespace).Get(ipAddress.Spec.ParentRef.Name)
+		if apierrors.IsNotFound(err) {
+			// cleaning all IPAddress without an owner reference IF the time since it was created is greater than 60 seconds (default timeout value on the kube-apiserver)
+			// This is required because during the Service creation there is a time that the IPAddress object exists but the Service is still being creaated
```
typo "creaated"

danwinship commented on pkg/registry/core/service/ipallocator/controller/repairip.go:
```diff
@@ -0,0 +1,320 @@
+
+	}
+
+	return nil
```
If you keep `ipaddress.kubernetes.io/ip-family` then you should probably check/repair that too.

Also, should we check that there are no `IPAddress` objects in the service CIDR ranges that _aren't_ managed by the service controller?

danwinship commented on pkg/registry/core/service/ipallocator/controller/repairip.go:
```diff
@@ -0,0 +1,320 @@
+	return nil
+}
+
+func newIPAddress(name string, ipFamily v1.IPFamily, svc *v1.Service) *networkingv1alpha1.IPAddress {
```
You define this (and `serviceToRef`) in two places... add a helper somewhere?

danwinship commented on pkg/registry/core/service/ipallocator/controller/repairip.go:
```diff
@@ -21,7 +21,6 @@ import (
 	"fmt"
```
it seems like `"move repair loop to controller"` ought to be squashed with `"add repair loop"`?

aojea replied:
yes, you are right

danwinship commented on test/integration/etcd/data.go:
```diff
@@ -241,6 +241,13 @@ func GetEtcdStorageDataForNamespace(namespace string) map[schema.GroupVersionRes
 		},
```
likewise, seems like "`integration etcd data`" ought to be squashed into "`add integration test`"

aojea replied:
the integration test commit adds tests just for this feature, the `integration etcd data`  is data that is used for another sets of tests, since usually those are different reviewers, I prefer to keep it in different commits

aojea commented on pkg/apis/networking/types.go:
```diff
@@ -699,3 +700,58 @@ type ClusterCIDRList struct {
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// IPAddress represents an IP used in a Kubernetes cluster, commonly associated to a Service ClusterIP.
```
yeah, excellent point, the historical context is that this was going to be only a Service IPAddress, it represents a Virtual IP that has some Pod behind ... but then Gateway and GAMMA expressed their interest, since it solves some of their problems of relying on Services ... sig-network can also benefit because it allows to "evolve" Services to be able express more complex networking setups without having to keep overloading the Service object and using Gateway API ...

 I will define it as `IPAddress represents a Virtual IP in the cluster that acts as a frontend for logical a set of Pods, used for representing the Service ClusterIP, but may be used for other APIs, like Gateway`

cc: @thockin @robscott

aojea commented on staging/src/k8s.io/api/networking/v1alpha1/well_known_labels.go:
```diff
@@ -0,0 +1,31 @@
+
+const (
+
+	// LabelIPAddressFamily is used to indicate the IP family of a Kubernetes IPAddress.
+	// This label simplify dual-stack client operations allowing to obtain the list of
+	// IP addresses filtered by family.
+	LabelIPAddressFamily = "ipaddress.kubernetes.io/ip-family"
```
This is a big performance improvement for clients, because it allows to filter by selector, and makes it super easier ...
it is a kubernetes.io label only, used by the kubernets controllers ... should we validate it? should we reconcile it and fix it if is wrong?

aojea replied:
I don't think we do that with the other controllers

aojea replied:
I can add validation in the repair loop and "repair" if necessary

smarterclayton replied:
If you need efficient selection, why isn't IPFamily a field with a field selector?  And the value is set based on the name at create time and immutable?

aojea replied:
to be honest, I just took the easiest path to iterate on that, I wasn't considering it at the beginning, but once I started to implement the allocators, it is super useful to select by family.
I will add a TODO to evaluate the need to move it to a field selector,

thockin replied:
Can field selectors be used in watch?  I have never done that...

We can revisit this and either make it a field selector or a "magic" label or something.  I agree we need to make some guarantees

sftim replied:
> Can field selectors be used in watch? I have never done that...

AIUI you can select on `metadata.name` to watch a collection for a specific named object. Quite common if I've understood things right.

thockin replied:
It seems metav1.LabelSelector can not spec field selectors either

thockin replied:
If we could put field selectors in a metav1.LabelSelector (I know, naming...) then the namespace magic label would not have been needed.

aojea commented on pkg/registry/core/service/ipallocator/ipallocator.go:
```diff
@@ -0,0 +1,480 @@
+	}
+
+	if prefix.Addr().Is6() && prefix.Bits() < 64 {
+		return nil, fmt.Errorf("maximum service CIDR is /64 , got /%d", prefix.Bits())
```
yeah, now I remember your PR, you are right we should go with shorted, the other words are misleading

aojea commented on pkg/registry/core/service/ipallocator/ipallocator.go:
```diff
@@ -0,0 +1,480 @@
+		family = api.IPv4Protocol
+		// Don't use the IPv4 network's broadcast address, but don't just
+		// Allocate() it - we don't ever want to be able to release it.
+		ipLast = ipLast.Prev()
```
wait for @thockin , he is challenging we should use the zero and broadcast address too ... I had to look it and he is right XD 

```
RFC 1878

*Subnet all zeroes and all ones excluded. (Obsolete)
*Host all zeroes and all ones excluded.   (Obsolete)
```

khenidak replied:
please don't - while in theory it will work (since the address is rewritten in proxy before the packet ever leave the host stack) it will confuse anybody coming at it from networking side

aojea replied:
😄

aojea commented on pkg/registry/core/service/ipallocator/ipallocator.go:
```diff
@@ -0,0 +1,480 @@
+// or has already been reserved.  ErrFull will be returned if there
+// are no addresses left.
+func (a *Allocator) Allocate(ip net.IP) error {
+	return a.AllocateService(nil, ip)
```
heh, now I understand one of the bugs I had , I need to fix this

aojea commented on pkg/registry/core/service/ipallocator/ipallocator.go:
```diff
@@ -0,0 +1,480 @@
+	if dryRun {
+		// Don't bother finding a free value. It's racy and not worth the
+		// effort to plumb any further.
+		return a.CIDR().IP, nil
```
this is an exact copy of the other allocator, I can't find the email discussion, but this is where we settle ... Services API allocators are a bit special in the apiserver

aojea replied:
that is racy, that is why Tim added this comment, the only moment you can be sure that the IP address is free is when you allocate it

aojea commented on pkg/registry/core/service/ipallocator/ipallocator.go:
```diff
@@ -0,0 +1,480 @@
+		// effort to plumb any further.
+		return a.CIDR().IP, nil
+	}
+	label := a.CIDR()
+	prefix, err := netip.ParsePrefix(label.String())
+	if err != nil {
+		return net.IP{}, err
```
yes, we should

aojea commented on pkg/registry/core/service/ipallocator/ipallocator.go:
```diff
@@ -0,0 +1,480 @@
+	firstAddress netip.Addr   // first IP address within the range
+	lastAddress  netip.Addr   // last IP address within the range
+	family       api.IPFamily // family is the IP family of this range
+	max          int64        // cap the total number of IPs available to maxInt64
```
yeah, I think I should be carrying on some of this inconsistencies in the code, I have to add a test or something ...

aojea replied:
> netutils.RangeSize to uint64

this is an incompatible change :/

aojea replied:
filed https://github.com/kubernetes/utils/pull/277 , but will keep a local function here to not depend on that change ... since this code is gated as alpha this will not be a problem

aojea commented on pkg/registry/core/service/ipallocator/ipallocator.go:
```diff
@@ -0,0 +1,480 @@
+	for i = 0; i < subrangeMax; i++ {
+		ip = ip.Next()
+		// roll over if it is greater than the last IP allocatable in the range
+		if ip.Compare(a.lastAddress) == 1 {
```
you have only Less() or Compare() https://pkg.go.dev/net/netip

khenidak commented on pkg/registry/networking/ipaddress/strategy.go:
```diff
@@ -0,0 +1,103 @@
+
+// NamespaceScoped returns false because all IPAddresses is cluster scoped.
+func (ipAddressStrategy) NamespaceScoped() bool {
+	return false
```
This is where i think we need to give it more thoughts. There is a lot of things that can use `IPAddress` either within the system or operator. Are we ok saying that anybody wants to use ipaddress object will have to use use it non-namespaced?

I have no opinion as of now, but i would like us to make an informed decision on that one.

aojea replied:
I've discussed this in the original KEP and the problem is cross-namespaced referencing, that seems to be not a thing ... or a complex thing at least

aojea replied:
... also the ClusterIP of a service is unique in a cluster, having IPAddress namespaced and apply the restriction that has to be unique  will be very confusing for users

khenidak replied:
ack.

khenidak commented on pkg/registry/core/service/ipallocator/controller/repairip.go:
```diff
@@ -0,0 +1,539 @@
+			r.recorder.Eventf(svc, nil, v1.EventTypeWarning, "ClusterIPNotValid", "ClusterIPValidation", "Cluster IP %s(%s) is of ip family that is no longer configured on cluster; please recreate Service", ip, family)
+			runtime.HandleError(fmt.Errorf("the ClusterIP [%v]: %s for Service %s/%s is of ip family that is no longer configured on cluster; please recreate Service", family, ip, svc.Name, svc.Namespace))
+			continue
+		} else if !network.Contains(ip) {
```
You don't need else-if here, logic reset and starts on top due to the `continue` at L339

aojea replied:
indeed

khenidak commented on pkg/registry/core/service/ipallocator/controller/repairip.go:
```diff
@@ -0,0 +1,539 @@
+		ipaddress, err := r.ipAddressLister.Get(ip.String())
+		if apierrors.IsNotFound(err) {
+			// ClusterIP doesn't seem to be allocated,
+			r.recorder.Eventf(svc, nil, v1.EventTypeWarning, "ClusterIPNotAllocated", "ClusterIPAllocation", "Cluster IP [%v]: %s is not allocated; repairing", family, ip)
```
This is a 3 states state-machine
1. Was not allocated: service ip was selected but have not correctly create the ipaddress object.
2. Fixed-Allocated: we managed to get the ip the service said it wants
3. Failed-Allocated: can't allocate the ip the service wants (only if we get duplicate error on create). 
4. Failed-general error: api-server was not cooperating at the time of create.

khenidak commented on pkg/registry/core/service/ipallocator/controller/repairip.go:
```diff
@@ -0,0 +1,539 @@
+			}
+			continue
+		}
+		if err != nil {
```
move this right after `ipaddress, err := r.ipAddressLister.Get(ip.String())`
Get the result 
if error && !IsNotFound (bail out, can't do anything now really)
if err && isNotFound(create or fail)
if no error && not owned by the service (own it?)

khenidak commented on pkg/registry/core/service/ipallocator/controller/repairip.go:
```diff
@@ -0,0 +1,539 @@
+		if !reflect.DeepEqual(serviceToRef(svc), ipaddress.Spec.ParentRef) {
+			r.recorder.Eventf(svc, nil, v1.EventTypeWarning, "ClusterIPNotAllocated", "ClusterIPAllocation", "the ClusterIP [%v]: %s for Service %s/%s has a wrong reference; repairing", family, ip, svc.Namespace, svc.Name)
+			runtime.HandleError(fmt.Errorf("the ClusterIP [%v]: %s for Service %s/%s has a wrong reference %v; repairing", family, ip, svc.Name, svc.Namespace, ipaddress.Spec.ParentRef))
+			_, err := r.client.NetworkingV1alpha1().IPAddresses().Update(context.Background(), newIPAddress(ip.String(), family, svc), metav1.UpdateOptions{})
```
What if it was owned by another service?

aojea replied:
oh, you are right, I should verify that doesn't exists Services with the same ClusterIP ... it can not happen ... unless there is a bug

khenidak commented on pkg/registry/core/service/ipallocator/controller/repairip.go:
```diff
@@ -0,0 +1,539 @@
+
+		if !verifyIPAddressLabels(ipaddress) {
+			runtime.HandleError(fmt.Errorf("the ClusterIP [%v]: %s for Service %s/%s has wrong labels %v; repairing", family, ip, svc.Name, svc.Namespace, ipaddress.Labels))
+			_, err := r.client.NetworkingV1alpha1().IPAddresses().Update(context.Background(), newIPAddress(ip.String(), family, svc), metav1.UpdateOptions{})
```
What if it has another label for say "another operator" created it?

khenidak commented on pkg/registry/core/service/ipallocator/controller/repairip.go:
```diff
@@ -0,0 +1,539 @@
+		r.recorder.Eventf(ipAddress, nil, v1.EventTypeWarning, "IPAddressNotAllocated", "IPAddressAllocation", "IPAddress %s appears to have been modified, not referencing a Service %v: cleaning up", ipAddress.Name, ipAddress.Spec.ParentRef)
+		err := r.client.NetworkingV1alpha1().IPAddresses().Delete(context.Background(), ipAddress.Name, metav1.DeleteOptions{})
+		if err != nil && !apierrors.IsNotFound(err) {
+			return err
```
is this an error really? what if another thing is updating it, say a delete service operation is executing side by side?

aojea replied:
the errors are retried, so it is safe, it will go again through the sync loop and reconcile correctly

khenidak commented on pkg/registry/core/service/ipallocator/controller/repairip.go:
```diff
@@ -0,0 +1,539 @@
+	onFirstSuccess()
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(r.ipWorker, r.workerLoopPeriod, stopCh)
```
Have we considered the ramification of having two loops operating on the same object graph? wouldn't it make more sense (and less racy) to work on services then IPs?

aojea replied:
The two loops are completely independent, Services loop: check that each clusterIP has an associated IPAddress and that this IPAddress is correct, if it doesn't exist or is wrong it repairs.

IPAddress loop: check that each IPAddress is associated to the corresponding Service, if the IP is orphan it deletes it, but it can only be orphan if the service is referencing doesn't exist

The added complexity has a great benefit that is performance, instead of working every X mins for doing the reconciliation, every event is processed as soons is received ... there is an integration test that goes from 3min to repair an IPAddress to ~5 sec with this

khenidak commented on pkg/registry/core/service/ipallocator/controller/repairip.go:
```diff
@@ -0,0 +1,539 @@
+	return nil
+
+}
+func newIPAddress(name string, ipFamily v1.IPFamily, svc *v1.Service) *networkingv1alpha1.IPAddress {
```
it would also make sense that this func also create the object and return the errors. There is a 100% chance that the object created by this function will be immediately posted to api-server after.

khenidak commented on pkg/registry/core/service/ipallocator/controller/repairip.go:
```diff
@@ -0,0 +1,539 @@
+			}
+		}
+		// requeue after the grace period
+		r.ipQueue.AddAfter(key, gracePeriod-ipLifetime)
```
`processNextWorkIp()` operates on each IP, calling  `handleIpErr` which re-Q in case of error. Why are we re-Qing here?

aojea replied:
when you create a Service PUT /service
1. internally you create an IP PUT /ipaddress
2. something happen here and the apiserver takes a bit of time to create the service
3. if the repair loop sees the IPAddress before the Service is created it will think that is orphan

to avoid any problem, I just queue to check it after the grace period, if the apiserver was slow the Service should exist, if the Service failed to be created or just crashed, this IP will be orphan

khenidak reviewed (commented):
More review comments. Slowly working my way to reach the finish line.

cici37 commented:
/remove-sig api-machinery

khenidak commented on pkg/apis/networking/types.go:
```diff
@@ -699,3 +700,56 @@ type ClusterCIDRList struct {
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// IPAddress represents a Virtual IP used in a Kubernetes cluster, that acts as a frontend for a
```
Modify to represent usage within core apis and other controls if they need to (but may be used for other APIs,
// like Gateway API.).

e.g.
`Each IPAddress object represents a single IP of a single IP Family. The object is designed to be used by apis that operate on IPs. The object is used by core apis for IPAddress allocation to service`

or something like that.

khenidak replied:
Modify to represent usage within core apis and other controls if they need to (remove the last part `but may be used for other APIs, like Gateway API.`).

e.g.
Each IPAddress object represents a single IP of a single IP Family. The object is designed to be used by apis that operate on IPs. The object is used by core apis for IPAddress allocation to service

or something like that.

khenidak commented on staging/src/k8s.io/api/networking/v1alpha1/register.go:
```diff
@@ -22,13 +22,18 @@ import (
 const GroupName = "networking.k8s.io"
 
-// SchemeGroupVersion is group version used to register these objects.
+// SchemeGroupVersion is group version used to register these objects
```
nit: `objects in this package.` (not `these objects`)

khenidak commented on staging/src/k8s.io/api/networking/v1alpha1/well_known_labels.go:
```diff
@@ -0,0 +1,31 @@
+	// LabelIPAddressFamily is used to indicate the IP family of a Kubernetes IPAddress.
+	// This label simplify dual-stack client operations allowing to obtain the list of
+	// IP addresses filtered by family.
+	LabelIPAddressFamily = "ipaddress.kubernetes.io/ip-family"
```
Why is this not part of the type?

aojea replied:
in order to use selectors

aojea replied:
and because is derived ... does it make sense?

khenidak commented on staging/src/k8s.io/api/networking/v1alpha1/types.go:
```diff
@@ -95,3 +96,69 @@ type ClusterCIDRList struct {
+
+// ParentReference describes a reference to a parent object.
+type ParentReference struct {
+	// Group is the group of the referent.
```
while the word `referent` is correct i doubt that it is clear specially for those who are not native english speaker.

aojea replied:
agree, I copied pasted :) . 
I've replaced it by `of the object being referenced.`

khenidak reviewed (commented):
Commit: Add IPAddress API
Hash: 43a2a35774d959ec241226dd61251f9ab6d5175e

khenidak reviewed (commented):
Commit: make update
hash:  5a1d6cbccd0ff711746220401e8529ef0b3aa2fe

it would be great to move this commit to the top of commit stack. My experience is these tend to get updated as you modify docs, apis, etc having it at the top of the commit makes it easier to just drop it and `make update` to generate it again.

khenidak reviewed (commented):
Commit: add feature gate

Hash: 5d26b3fa1651e922b49e3c23ef680d6f4a0deb8f
Move to the bottom of the commit stack (before apis)

Usually those never change once created as a commit, so the safest place is the first entry in the commit stack.

khenidak commented on pkg/registry/core/service/ipallocator/ipallocator.go:
```diff
@@ -0,0 +1,480 @@
+}
+
+// ForEach executes the function in each allocated IP
+func (a *Allocator) ForEach(f func(net.IP)) {
```
why do we need this func?

aojea replied:
It is used by the Repair loop on the bitmap allocator.
This new allocator don't really need it but it needs to match the Interface so we can switch the old one with this

khenidak commented on pkg/registry/core/service/ipallocator/ipallocator.go:
```diff
@@ -0,0 +1,480 @@
+
+const ControllerName = "ipallocator.k8s.io"
+
+// Allocator implements current ipallocator interface using IPAddress API object
```
General Thoughts: 
My biggest concern is on the way we implemented this is it is o(N) where n is the # of IPs in the systems. That is quite something giving the fact that some systems will run with thousands if not 10s of thousands of IPs. the start from the top and from the bottom of the range helps but i am not sure of how good it is.

The second biggest concern I have is the idea that the expected stdev perf for a set of allocation operations is just too high. Think of the delta between  the fastest allocation (lucky to find a free ip early in the loop) or slowest (where we need to iterate over the entire list of IPs). 

if we agree that these are concern. Can we start by considering alternative options specifically replacing the iteration with a process that does not need to iterate over everything.

aojea replied:
fully agree, and I put a lot of thought on that and is also mentioned in the KEP, I also added benchmarks in both allocators to compare, see pkg/registry/core/service/ipallocator/ipallocator_test.go. and added it as a part of the beta criteria ;)

> Improve performance on the allocation logic, O(1) for allocating a free address

https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/1880-multiple-service-cidrs#beta

But I think that for alpha this is good enough

khenidak commented on pkg/registry/core/service/storage/alloc.go:
```diff
@@ -399,7 +402,20 @@ func (al *Allocators) allocIPs(service *api.Service, toAlloc map[api.IPFamily]st
+			if utilfeature.DefaultFeatureGate.Enabled(features.MultiCIDRServiceAllocator) {
+				// if dryRun is being used the allocator is implemented by ipallocator.dryRunRange
+				// TODO: modify the Interface for the ipallocator
+				svcAllocator, ok := allocator.(*ipallocator.Allocator)
```
L400 already does the get allocator by family.  if you modify 
`serviceIPAllocatorsByFamily[...]`
to 
`func getServiceAllocatorByFamily(...) allocator` then
you won't need  `svcAllocator, ok := allocator.(*ipallocator.Allocator)`  nor `the feature gate`.

`func getServiceAllocatorByFamily(...) allocator` can return an allocator that either calls old/new and move all functions to pass in a service ref.

aojea replied:
I think that is much better, and we should move to that in beta, but at this stage I think that having it this way will be easier to debug and to understand , despite is much verbose ...

aojea replied:
I misunderstood, I see that way is simpler

aojea replied:
However, this will need another change when we move to multiple ServiecCIDRs https://github.com/kubernetes/kubernetes/pull/116516/commits/6f60e274e82f5ba5a94f12819ca3cc98fc352cb2 so if you don't mind I will not update it now

khenidak replied:
ack - I will add it to my list for cidrs PR.

khenidak commented on pkg/registry/core/service/storage/alloc.go:
```diff
@@ -409,7 +425,20 @@ func (al *Allocators) allocIPs(service *api.Service, toAlloc map[api.IPFamily]st
-			if err := allocator.Allocate(parsedIP); err != nil {
+			var err error
+			if utilfeature.DefaultFeatureGate.Enabled(features.MultiCIDRServiceAllocator) {
+				// if dryRun is being used the allocator is implemented by ipallocator.dryRunRange
```
we have an open discussion to `dryrun` implementation in the new allocator. I think this proves the point that we need to do it anyway, since eventually we want to remove the old allocator (and thus we will have to implement the dry run anyway).

khenidak reviewed (commented):
Commit: allow to set the service reference on the allocator
Hash: a16f5ffa9700d576c70cfc391bc53b11726bc442

aojea commented:
> Commit: make update hash: [5a1d6cb](https://github.com/kubernetes/kubernetes/commit/5a1d6cbccd0ff711746220401e8529ef0b3aa2fe)
> 
> it would be great to move this commit to the top of commit stack. My experience is these tend to get updated as you modify docs, apis, etc having it at the top of the commit makes it easier to just drop it and `make update` to generate it again.

I need it there to work on the intermediate commits, my workflow may be different but I use `git rebase -i`and stops on the commit I want to modify. If I don't have the generated code at that point then all tests will fail

aojea commented on pkg/registry/core/service/ipallocator/ipallocator.go:
```diff
@@ -0,0 +1,480 @@
+// NewIPAllocator returns an IP allocator that use the IPAddress object
+// to track the assigned IP addresses, using an informer cache as storage.
+func NewIPAllocator(
+	cidr *net.IPNet,
```
I've started with that but the PR grew insane and it was super complex to review, this allows us to do an intermediate step that is fully functional with 1-1 replacement ... the next PR will be only about this multi-cidr allocator

aojea commented on pkg/registry/core/service/ipallocator/ipallocator.go:
```diff
@@ -0,0 +1,480 @@
+	}
+
+	// small CIDRs may return a negative number
+	if max < 0 {
```
if you use a /32 or /128 😄

aojea commented on pkg/registry/core/service/ipallocator/ipallocator.go:
```diff
@@ -0,0 +1,480 @@
+		net:             cidr,
+		firstAddress:    ipFirst,
+		lastAddress:     ipLast,
+		rand:            rand.New(rand.NewSource(time.Now().UnixNano())),
```
```
	// Typically a non-fixed seed should be used, such as time.Now().UnixNano().
	// Using a fixed seed will produce the same output on every run.
	r := rand.New(rand.NewSource(99))
```

https://pkg.go.dev/math/rand

danwinship replied:
yes, I mean, we do this in a centralized location: https://github.com/kubernetes/kubernetes/blob/v1.26.2/staging/src/k8s.io/component-base/cli/run.go#L89

aojea replied:
oh, I didn't know about that, but I think that this is a nit, right?

aojea commented on pkg/registry/core/service/ipallocator/ipallocator.go:
```diff
@@ -0,0 +1,480 @@
+	if err != nil {
+		return net.IP{}, err
+	}
+	if a.Free() == 0 {
```
it depends on `ipAddressSynced` to be true and for dryRun() the conventions is to not check the allocator logic https://github.com/kubernetes/kubernetes/pull/115075/commits/f35fb15be32671edf47e32eb1fb528b0564eccc9#r1129923931. (I know you don't like it though) :)

aojea commented on pkg/registry/core/service/ipallocator/ipallocator.go:
```diff
@@ -0,0 +1,480 @@
+	// try to get a value from the upper range [a.offset, max]
+	start := a.rand.Int63n(subrangeMax) + int64(a.offset)
+	ip := getIndexedAddr(prefix, start)
+	offsetAddress := getIndexedAddr(prefix, int64(a.offset))
```
it is used later for the subrange iteration, since we iterate first between (offset, max) and later from (1,offset)

aojea replied:
but this is related to https://github.com/kubernetes/kubernetes/pull/115075/commits/f35fb15be32671edf47e32eb1fb528b0564eccc9#r1099348813, so I will do both suggestions

aojea commented on pkg/registry/core/service/ipallocator/ipallocator.go:
```diff
@@ -0,0 +1,480 @@
+		}
+		name := ip.String()
+		_, err = a.ipAddressLister.Get(name)
+		if !apierrors.IsNotFound(err) {
```
yep

aojea commented on pkg/registry/core/service/ipallocator/ipallocator.go:
```diff
@@ -0,0 +1,480 @@
+	}
+	name := ip.String()
+	if _, err := a.ipAddressLister.Get(name); err != nil {
+		return nil
```
yep

aojea commented on pkg/registry/core/service/ipallocator/ipallocator.go:
```diff
@@ -0,0 +1,480 @@
+		return
+	}
+	for _, ip := range ips {
+		f(netutils.ParseIPSloppy(ip.Name))
```
yeah, but is hard to implement that in the linters, I have it in my TODO to narrow the linter for net.ParseIP to only the places that validate the APIs, but these files can import helpers from all the repo.

I was expecting to do this transition using netip.Addr, so eventually we'll have netip.Addr for the new code, and only  ParseIPSloppy for the API validation

aojea commented on pkg/registry/core/service/ipallocator/ipallocator.go:
```diff
@@ -0,0 +1,480 @@
+}
+
+// broadcastAddress returns the broadcast address of the subnet
+func broadcastAddress(subnet netip.Prefix) netip.Addr {
```
is a one shot operation in the allocator, we don't need to optimize much here

aojea commented on pkg/registry/core/service/ipallocator/ipallocator.go:
```diff
@@ -0,0 +1,480 @@
+			// update metrics
+			a.metrics.incrementAllocationErrors(label.String(), "static")
+		}
+		return &ErrNotInRange{ip, label.String()}
```
fixed

aojea commented on pkg/registry/core/service/ipallocator/ipallocator.go:
```diff
@@ -0,0 +1,480 @@
+
+// AllocateNext return an IP address that wasn't allocated yet
+func (a *Allocator) AllocateNext() (net.IP, error) {
+	return a.AllocateNextService(nil)
```
added a comment that this is just for testing

aojea commented on pkg/registry/core/service/ipallocator/ipallocator.go:
```diff
@@ -0,0 +1,480 @@
+	}
+
+	// rand.Int63n panics for values <= 0
+	if a.offset == 0 {
```
done

aojea committed:
make update

Change-Id: I19e12ca05d977dca63043cb07ecf8a90e0e525c5

aojea committed:
plumb new ipallocators in the apiserver

plumb the new allocators from the control-plane instance to the
registry_core

Change-Id: I240cc91942260f725492597f1b599f7480560b1e

aojea committed:
clusterip allocator based on IPAddress API

add a new ClusterIP allocator that uses the new IPAddress API resource
and an informer as the backend, instead a bitmap snapshotted on etcd.

Change-Id: Ia891a2900acd2682d4d169abab65cdd9270a8445

aojea commented:
@khenidak @danwinship @thockin updated addressing comment and fixing some embarrassing bugs

This is the 1st phase, it introduce IPAddress object and an allocator and repair loop using the new object
The [2nd phase introduces a new ServiceCIDR object](https://github.com/kubernetes/kubernetes/pull/116516), makes the new allocator and repair loop added in phase 1 to work with it. 

I'd like to (please) have this 2 phases before code freeze, I know that is a short time. 

Phase 1 and 2 togethers allow to have a full complete experience of multiple ServiceCIDR, but we still have to deal with the ClusterCIDR relation.

There is a 3rd phase that is how to manage this relation with ClusterCIDRs, but since both features are alpha, I think that is reasonable to work on those at the same time in the next cycle

aojea commented:
```
Kubernetes e2e suite: [It] [sig-network] HostPort validates that there is no conflict between pods with same hostPort but different hostIP and protocol [LinuxOnly] [Conformance] expand_less	27s
{ failed [FAILED] wait for pod pod1 timeout, err:Told to stop trying after 2.206s.
The phase of Pod pod1 is Failed which is unexpected.
In [It] at: test/e2e/network/hostport.go:219 @ 03/13/23 22:56:38.82
}
```

unrelated but weird error


/test pull-kubernetes-e2e-kind-ipv6
--


<br class="Apple-interchange-newline">

pacoxu commented:
/remove-area kubeadm

aojea commented:
/test pull-kubernetes-e2e-gce-cos-alpha-features

it should be fixed now

smarterclayton commented on staging/src/k8s.io/api/networking/v1alpha1/well_known_labels.go:
```diff
@@ -0,0 +1,31 @@
+	// objects to be managed by different controllers or entities within the
+	// same cluster. It is highly recommended to configure this label for all
+	// IPAddress objects.
+	LabelManagedBy = "ipaddress.kubernetes.io/managed-by"
```
Why isn't this is the controller / ownerref?

aojea replied:
NonNamespaces objects can not be referenced by namespaced objects like Services

> In v1.20+, if the garbage collector detects an invalid cross-namespace ownerReference, or a cluster-scoped dependent with an ownerReference referencing a namespaced kind, 

or is another kind of reference?

thockin replied:
We need to fix owner ref.  Grr....

sftim replied:
If we can't fix owner reference, or find an equivalent mechanism that doesn't use labels, could the label key be `kubernetes.io/managed-by`? I'd like to just have one notion of things managing things rather than _n_ of them.

aojea replied:
that is a good point, we should open this thread with apimachinery about owner references

aojea replied:
@liggitt can you enlighten us if this is namspace -> non-namespace owneref something that can be fixed, is just nobody worked on it, or is something you don't see we can change it ... I'm willing to spend time on it if this can be fixed since will allow to remove a considerable amount of code from the repair loop for clusterIPs

Ref https://github.com/kubernetes/kubernetes/pull/92743

liggitt replied:
both of the following were intentionally out of scope for owner references:
* cluster-scoped things owned by namespaced things (higher permission thing owned by something in a lower permission scope)
* namespaced things owned by things in other namespaces

both of those lines are permission boundaries, and letting arbitrary objects cross those lines was considered confusing (spooky action at a distance) and/or dangerous (especially with aspects of ownerReferences like blockOwnerDeletion)

see the note at https://kubernetes.io/docs/concepts/overview/working-with-objects/owners-dependents/#owner-references-in-object-specifications

aojea commented on staging/src/k8s.io/api/networking/v1alpha1/types.go:
```diff
@@ -95,3 +96,68 @@ type ClusterCIDRList struct {
+// the name of the object is the IP address in canonical format, four decimal digits separated
+// by dots suppressing leading zeros for IPv4 and the representation defined by RFC 5952 for IPv6.
+// Valid: 192.168.1.5 or 2001:db8::1 or 2001:db8:aaaa:bbbb:cccc:dddd:eeee:1
+// Invalid: 10.01.2.3 or 2001:db8:0:0:0::1
```
yes, this is specific for IP addresses, we have `ServiceCIDR` and `ClusterCIDR`objects to handle the CIDR needs

fedebongio commented:
/remove-sig api-machinery

khenidak reviewed (commented):
/lgtm
/approve

Thanks for putting all that effort @aojea !

khenidak commented:
by now i think this should be mandatory in sig-network PRs that looks like that.. inspired by @thockin 

![image](https://user-images.githubusercontent.com/10683203/225128816-53ac353e-dd55-4f0e-8ede-67eaf1f0c1c6.png)

aojea commented:
/test pull-kubernetes-e2e-gce

sisg storage tests 👀

aojea commented:
> PR needs rebase.

rebased

@thockin we need `sudo` 😄 

> /lgtm
> /approve

aojea commented:
> Kubernetes e2e suite: [It] [sig-cli] Kubectl client kubectl subresource flag GET on status subresource of built-in type (node) returns identical info as GET on the built-in type expand_less	

known flake 

/test pull-kubernetes-e2e-kind
https://github.com/kubernetes/kubernetes/pull/116619

thockin commented on pkg/apis/networking/types.go:
```diff
@@ -699,3 +700,55 @@ type ClusterCIDRList struct {
+	Name string
+	// UID is the uid of the object being referenced.
+	// +optional
+	UID types.UID
```
Do we really need UID?

aojea replied:
I think this came up during the KEP review and in paranoid mode, but I think that is excessive, at least for Services the ClusterIP -> IPAddress -> ParentRef -> Service -> ClusterIP is able to build a directed graph ... the solution on the repair will be just to update the UID that doesn't sound necessary ... will remove it in next iteration

aojea replied:
dropped

thockin commented on pkg/apis/networking/validation/validation.go:
```diff
@@ -741,3 +742,77 @@ func validateClusterCIDRUpdateSpec(update, old *networking.ClusterCIDRSpec, fldP
+	if err != nil {
+		errs = append(errs, err.Error())
+	} else if ip.String() != name {
+		errs = append(errs, "not a valid ip in canonical format")
```
"must be a canonical format IP address"

thockin reviewed (commented):
On behalf of the review done by others and the fact that it is alpha

/lgtm
/approve

thockin commented:
![Uploading indiana_jones_hat_under_the_door.gif…]()

thockin commented:
/retest

sftim commented:
Changelog suggestion
```markdown
- Added a new IPAddress object kind
- Added a new ClusterIP allocator. The new allocator removes previous Service CIDR block size limitations for IPv4, and limits IPv6 size to a /64
```

something like that?

sftim commented on staging/src/k8s.io/api/networking/v1alpha1/well_known_labels.go:
```diff
@@ -0,0 +1,33 @@
+	// This label simplify dual-stack client operations allowing to obtain the list of
+	// IP addresses filtered by family.
+	LabelIPAddressFamily = "ipaddress.kubernetes.io/ip-family"
+	// LabelManagedBy is used to indicate the controller or entity that manages
+	// an IPAddress. This label aims to enable different IPAddress
+	// objects to be managed by different controllers or entities within the
+	// same cluster. It is highly recommended to configure this label for all
+	// IPAddress objects.
```
We should register this - see https://kubernetes.io/docs/reference/labels-annotations-taints/

sftim commented on staging/src/k8s.io/api/networking/v1alpha1/well_known_labels.go:
```diff
@@ -0,0 +1,33 @@
+
+const (
+
+	// TODO: Use IPFamily as field with a field selector,And the value is set based on
+	// the name at create time and immutable.
+	// LabelIPAddressFamily is used to indicate the IP family of a Kubernetes IPAddress.
+	// This label simplify dual-stack client operations allowing to obtain the list of
+	// IP addresses filtered by family.
```
If we need a label, we should register this - see https://kubernetes.io/docs/reference/labels-annotations-taints/

I'd go for `network.kubernetes.io/ip-address-family`. Network is a nice general concept whereas IP address feels very specific. And, as a project, we control the namespace fairly effectively.

aojea commented:
/milestone v1.27

approved before code freeze 
https://github.com/kubernetes/kubernetes/pull/115075#issuecomment-1469024136

aojea commented:
> On behalf of the review done by others and the fact that it is alpha
> 
> /lgtm /approve

@thockin it seems it didn't pick the approve label correctly, can you retag?

thockin commented:
/approve

BenTheElder commented:
Congrats @aojea :-)

aojea commented:
Summarizing issues from this PR to not forget, please add the things you find missing, so I can start incorporating them to the follow up in https://github.com/kubernetes/kubernetes/pull/116516

- [ ] Document upgrades and rollbacks, specially for those clusters with size that are now allowed that were not allowed before
- [x] Owner references, namespaces -> non-namespaces, it seems we should keep using label https://github.com/kubernetes/kubernetes/pull/115075#discussion_r1136876396
- [ ] Field-selector vs Label-selector, the flexibility of labels makes it simpler to create controllers that handle IPAddresses and need to be dual-stack https://github.com/kubernetes/kubernetes/pull/115075#discussion_r1136871581, I prefer to keep using labels https://github.com/kubernetes/client-go/issues/604 since I expect most of the consumers will need to do a lot of filtering client side
- [x] Drop ParentReference UID
- [ ] If we use labels, document them properly

cici37 commented:
/triage accepted
